docs/tutorials/basic/data.md

```diff
@@ -187,22 +116,302 @@ class SimpleIter:
+net = mx.sym.SoftmaxOutput(data=net, name='softmax')
+print(net.list_arguments())
+print(net.list_outputs())
+```
+
+Here, there are four variables that are learnable parameters:
+the *weights* and *biases* of FullyConnected layers *fc1* and *fc2*,
+two variables for input data: *data* for the training examples
+and *softmax_label* contains the respective labels and the *softmax_output*.
+
+The *data* variables are called free variables in MXNet's Symbol API.
+To execute a Symbol, they need to bound with data.
+[Click here learn more about Symbol](http://mxnet.io/tutorials/basic/symbol.html).
+
+We use the data iterator to feed the neural networks via MXNet's `module` API.
+[Click here to learn more about Module](http://mxnet.io/tutorials/basic/module.html).
+
+
+```python
+import logging
+logging.basicConfig(level=logging.INFO)
+
+n = 32
+data_iter = SimpleIter(['data'], [(n, 100)],
+                  [lambda s: np.random.uniform(-1, 1, s)],
+                  ['softmax_label'], [(n,)],
+                  [lambda s: np.random.randint(0, num_classes, s)])
+
+mod = mx.mod.Module(symbol=net)
+mod.fit(data_iter, num_epoch=5)
+```
+
+## Record IO
+Record IO is a file format used by MXNet for data IO.
+It compactly packs the data for efficient read
+and writes from various filesystems
+including distributed file systems like Hadoop HDFS and AWS S3.
+You can learn more about the thinking behind the design of `RecordIO`
+[here](http://mxnet.io/architecture/note_data_loading.html).
+
+MXNet provides __`MXRecordIO`__ and __`MXIndexedRecordIO`__
+for sequential access of data and random access of the data.
+
+### MXRecordIO
+First let's look at an example on how to read and write sequentially
+using `mx.recordio.MXRecordIO`. The files are named with a `.rec` extension.
+
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'w')
+for i in range(5):
+    record.write('record_%d'%i)
+record.close()
+```
+
+We can read the data back by opening the file with a option `r` as below:
+
+```python
+record = mx.recordio.MXRecordIO('tmp.rec', 'r')
+while True:
+    item = record.read()
+    if not item:
+        break
+    print (item)
+record.close()
+```
+
+
+### MXIndexedRecordIO
+`MXIndexedRecordIO` supports random or indexed access to the data.
+We will create a indexed record file and a corresponding index file as below:
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'w')
+for i in range(5):
+    record.write_idx(i, 'record_%d'%i)
+record.close()
+```
+
+Now we can access the individual records using the keys
+
+
+```python
+record = mx.recordio.MXIndexedRecordIO('tmp.idx', 'tmp.rec', 'r')
+record.read_idx(3)
+```
+
+You can also list all the keys in the file.
+
+
+```python
+record.keys
+```
 
+### Packing and Unpacking data
+
+Each record in a .rec file can contain arbitrary binary data,
+however most deep learning tasks require that data arrives a label/data format.
```

Comment:
- however -> However,
- require data arrives in a label/data format -> require data to be input in label/data format.